nft-client/src/cts1.rs

```diff
@@ -0,0 +1,377 @@
+        }
+        Ok(())
+    }
+}
+
+/// Additional data bytes included with transfers the parameter for the CTS1
+/// contract function "transfer".
+#[derive(Debug, Clone)]
+pub struct ReceiveHookData(pub Vec<u8>);
+
+/// Serialization for the additional data, serialized as according to the CTS1
+/// specification.
+impl Serial for ReceiveHookData {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u16::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        for byte in &self.0 {
+            byte.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// Parse the additional data from a hex string.
+impl FromStr for ReceiveHookData {
+    type Err = hex::FromHexError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(ReceiveHookData(hex::decode(s)?.to_vec()))
+    }
+}
+
+/// A description of a transfer as according to the CTS1 specification.
+#[derive(Debug)]
+pub struct Transfer {
+    pub token_id:     TokenIdVec,
+    pub amount:       u64,
+    pub from:         Address,
+    pub to:           Address,
+    pub receive_name: Option<OwnedReceiveName>,
+    pub data:         ReceiveHookData,
+}
+
+/// Serialization of a transfer, as according to the CTS1 specification.
+impl Serial for Transfer {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        self.token_id.serial(out)?;
+        self.amount.serial(out)?;
+        self.from.serial(out)?;
+        self.to.serial(out)?;
+        if let Address::Contract(_) = self.to {
+            self.receive_name
+                .as_ref()
+                .ok_or_else(W::Err::default)?
+                .serial(out)?;
+        }
+        self.data.serial(out)?;
+        Ok(())
+    }
+}
+
+/// The parameter type for the NFT contract function `CTS1-NFT.transfer`.
+#[derive(Debug)]
+pub struct TransferParams(pub Vec<Transfer>);
+
+/// Serialization of the transfer parameter, as according to the CTS1
+/// specification.
+impl Serial for TransferParams {
+    fn serial<W: Write>(&self, out: &mut W) -> Result<(), W::Err> {
+        let len = u8::try_from(self.0.len()).map_err(|_| W::Err::default())?;
+        len.serial(out)?;
+        for transfer in &self.0 {
+            transfer.serial(out)?;
+        }
+        Ok(())
+    }
+}
+
+/// The type of update for an operator update.
+pub enum OperatorUpdate {
+    /// Remove the operator.
+    Remove,
+    /// Add an address as an operator.
+    Add,
+}
+
+/// The deserialization of an operator update as according to the CTS1
+/// specification.
+impl Deserial for OperatorUpdate {
+    fn deserial<R: Read>(source: &mut R) -> Result<Self, ParseError> {
+        let discriminant = source.read_u8()?;
+        match discriminant {
+            0 => Ok(OperatorUpdate::Remove),
+            1 => Ok(OperatorUpdate::Add),
+            _ => Err(ParseError::default()),
+        }
+    }
+}
+
+/// A SHA256 digest.
+pub type Sha256 = [u8; 32];
```

Comment:
Why are all these types just type aliases? This is not in-line with the strongly typed nature of everything else.

The SDK already provides a Hash type (with a phantom marker to distinguish uses), why can you not use that?

Reply:
The hope is to replace this module with the concordium-cts1 crate, which will not depend on concordium-rust-sdk.

Another minor issue is that HashBytes from concordium-rust-sdk does not implement the Serial trait from concordium_contracts_common.